zerotrace.go

```diff
@@ -15,223 +15,221 @@ import (
+			if state.isFinished() {
+				return state.calcRTT()
 			}
 		}
 	}
+}
+
+// sendTracePkts sends a burst of trace packets to our target.  Once a packet
+// was sent, it's written to the given channel.
+func (z *ZeroTrace) sendTracePkts(
+	c chan *tracePkt,
+	conn net.Conn,
+	wg *sync.WaitGroup,
+) {
+	defer wg.Done()
+
+	dstAddr, err := extractRemoteIP(conn)
+	if err != nil {
+		l.Printf("Error extracting remote IP address from connection: %v", err)
+		return
+	}
+	pktPayload, err := createPkt(conn)
+	if err != nil {
+		l.Printf("Error creating trace packet payload: %v", err)
+		return
+	}
 
-	return state.CalcRTT(), nil
+	start := time.Now().UTC()
+	defer func() {
+		diff := time.Now().UTC().Sub(start)
+		l.Printf("Sent trace packets in: %v", diff)
+	}()
+	for ttl := z.cfg.TTLStart; ttl <= z.cfg.TTLEnd; ttl++ {
+		// Parallelize the sending of trace packets.
+		go func(ttl int) {
+			hdr := newIpv4Header(ttl, 0, dstAddr, len(pktPayload))
+			// Send n probe packets for redundancy, in case some get lost.
+			// Each probe packet shares a TTL but has a unique ID.
+			for n := 0; n < z.cfg.NumProbes; n++ {
+				ipID, err := z.ipids.borrow()
+				if err != nil {
+					l.Printf("Error borrowing IPID: %v", err)
+					continue
+				}
+				hdr.ID = int(ipID)
+				if err = z.rawConn.WriteTo(hdr, pktPayload, nil); err != nil {
+					l.Printf("Error sending trace packet: %v", err)
+					continue
+				}
+				c <- &tracePkt{
+					ttl:  uint8(ttl),
+					ipID: ipID,
+					sent: time.Now().UTC(),
+				}
+			}
+		}(ttl)
+	}
 }
 
-// recvRespPkts uses the given pcap handle to read incoming packets and filters
-// for ICMP TTL exceeded packets that are then sent to the given channel.  The
-// function returns when the given quit channel is closed.
-func (z *ZeroTrace) recvRespPkts(pcapHdl *pcap.Handle, c chan *respPkt, quit chan bool) {
-	packetStream := gopacket.NewPacketSource(pcapHdl, pcapHdl.LinkType())
+// listen opens a pcap handle and begins listening for incoming ICMP packets.
+// New traceroutes register themselves with this function's event loop to
+// receive a copy of newly-captured ICMP packets.
+func (z *ZeroTrace) listen(pktStream chan gopacket.Packet) {
+	var (
+		ticker    = time.NewTicker(3 * time.Second)
+		receivers = make(map[receiver]bool)
+	)
 
+	l.Println("Starting listening loop.")
+	defer l.Println("Leaving listening loop.")
 	for {
 		select {
-		case <-quit:
-			l.Println("Done reading packets.")
+		case <-z.quit:
 			return
-		case pkt := <-packetStream.Packets():
-			if pkt == nil {
-				continue
-			}
-			ipLayer := pkt.Layer(layers.LayerTypeIPv4)
-			icmpLayer := pkt.Layer(layers.LayerTypeICMPv4)
-
-			if ipLayer == nil || icmpLayer == nil {
-				continue
-			}
-
-			// If it is an ICMP packet, check if it is the ICMP TTL
-			// exceeded one we are looking for
-			respPkt, err := z.extractRcvdPkt(pkt)
+		case <-ticker.C:
+			z.ipids.releaseUnanswered()
+			l.Printf("Released un-answered IP IDs; %d left.", z.ipids.size())
```

Comment:
Possibly redundant with the log message inside `releaseUnanswered`?

Reply:
Yes, fixed in c6224beeecb951ba338406dfeb9877631b665214.